### ISSUE TITLE
Eel & Tkinter: Localhost refuses connection

### ISSUE THREAD

**Issue by CentralHeating** [samuelhwilliams/Eel#167]
Hi, I have a button in a tkinter application which pops up an Eel window (chrome-app) with a page in it. Initially it showed the html page that was a static page (with javascript in it) in the same folder as the Tkinter app (the main folder containing most stuff in the one place to avoid the old path-to-file issues). Then I tried to get it to show another html, also in the main folder with the Tkinter script, but now I'm getting 'localhost connection refused' (ERR_CONNECTION_REFUSED). Same story when putting in an address on the web, it doesn't show the page in the Eel/Chrome-app browser.

Now when I try to show the original file it doesn't show any more either. I'm wondering if:
a) the reason it showed and now doesn't show is that after first running it, it hasn't shut down in some way and is blocking further runs of the script.
b) there's something to do with the threading that's causing issues. When the button in the Tkinter app is pressed (to make the Eel/Chrome-app browser window pop up), the button runs a command which starts a thread (using the Threading module) which then targets (i.e. calls/runs) another function which runs the Eel stuff. Like so:

def startthreadeel():
    teel = threading.Thread(target=runeel, name='threadeel')
    teel.start()
    teel.join()

def runeel():
    eel.init("")
    eel.start('leaflet.html', block=False)

Threading is necessary to stop the app from freezing up while any other events (like getting stuff from the web) are done/things are being processed. I'm at a loss as to why it worked and now doesn't and really can't find information to untangle this and resolve it. I'd appreciate any insight, thanks.

**Comment by CentralHeating** [comment]
UPDATE:

I ran it from the command line using the following:
import eel
eel.init("")
eel.start('leaflet.html', block=False)

This also popped up a window that said 'localhost connection refused'. I then took out the 'block=False' and ran the same line again:
eel.start('leaflet.html')

Now it worked, it seems to have unblocked the page. The command line gave the following traceback when doing this:

---------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "src\gevent\greenlet.py", line 766, in gevent._greenlet.Greenlet.run
  File "C:\Python27\lib\site-packages\eel\__init__.py", line 127, in run_lambda
    quiet=True)
  File "C:\Python27\lib\site-packages\bottle.py", line 3129, in run
    server.run(app)
  File "C:\Python27\lib\site-packages\bottle_websocket\server.py", line 17, in run
    server.serve_forever()
  File "C:\Python27\lib\site-packages\gevent\baseserver.py", line 367, in serve_forever
    self.start()
  File "C:\Python27\lib\site-packages\gevent\baseserver.py", line 305, in start
    self.init_socket()
  File "C:\Python27\lib\site-packages\gevent\pywsgi.py", line 1490, in init_socket
    StreamServer.init_socket(self)
  File "C:\Python27\lib\site-packages\gevent\server.py", line 146, in init_socket
    self.socket = self.get_listener(self.address, self.backlog, self.family)
  File "C:\Python27\lib\site-packages\gevent\server.py", line 157, in get_listener
    return _tcp_listener(address, backlog=backlog, reuse_addr=cls.reuse_addr, family=family)
  File "C:\Python27\lib\site-packages\gevent\server.py", line 256, in _tcp_listener
    sock.bind(address)
  File "<string>", line 1, in bind
error: [Errno 10048] Only one usage of each socket address (protocol/network address/port) is normally permitt
ed: ('localhost', 8000)
2019-07-10T21:39:34Z <Greenlet at 0x30812d8: run_lambda> failed with error
---------------------------------------------------------------------------------------------------

So it appears to be a port issue. In my original flailing trying to find an answer I also added options to the eel.start() function like so:

my_options = {
    'mode': "chrome-app", #or "chrome-app",
    'host': 'localhost',
    'host': '127.0.0.1',
    'port': 8080
}

eel.start('leaflet.html', options=my_options, block=False)

So it seems to have unblocked for the time being. I'm not sure what I'm looking at or what has occurred here, my intention is to package this app and distribute it so I can only imagine this issue will recur in quite a few other instances. Again any insight is appreciated.

**Comment by CentralHeating** [comment]
UPDATE:

I ran it from the command line using the following:
import eel
eel.init("")
eel.start('leaflet.html', block=False)

This also popped up a window that said 'localhost connection refused'. I then took out the 'block=False' and ran the same line again:
eel.start('leaflet.html')

Now it worked, it seems to have unblocked the page. The command line gave the following traceback when doing this:
Traceback (most recent call last):
File "src\gevent\greenlet.py", line 766, in gevent.greenlet.Greenlet.run
File "C:\Python27\lib\site-packages\eel_init.py", line 127, in run_lambda
quiet=True)
File "C:\Python27\lib\site-packages\bottle.py", line 3129, in run
server.run(app)
File "C:\Python27\lib\site-packages\bottle_websocket\server.py", line 17, in run
server.serve_forever()
File "C:\Python27\lib\site-packages\gevent\baseserver.py", line 367, in serve_forever
self.start()
File "C:\Python27\lib\site-packages\gevent\baseserver.py", line 305, in start
self.init_socket()
File "C:\Python27\lib\site-packages\gevent\pywsgi.py", line 1490, in init_socket
StreamServer.init_socket(self)
File "C:\Python27\lib\site-packages\gevent\server.py", line 146, in init_socket
self.socket = self.get_listener(self.address, self.backlog, self.family)
File "C:\Python27\lib\site-packages\gevent\server.py", line 157, in get_listener
return _tcp_listener(address, backlog=backlog, reuse_addr=cls.reuse_addr, family=family)
File "C:\Python27\lib\site-packages\gevent\server.py", line 256, in _tcp_listener
sock.bind(address)
File "", line 1, in bind
error: [Errno 10048] Only one usage of each socket address (protocol/network address/port) is normally permitt
ed: ('localhost', 8000)
2019-07-10T21:39:34Z <Greenlet at 0x30812d8: run_lambda> failed with error

So it appears to be a port issue. In my original flailing trying to find an answer I also added options to the eel.start() function like so:

my_options = {
'mode': "chrome-app", #or "chrome-app",
'host': 'localhost',
'host': '127.0.0.1',
'port': 8080
}

eel.start('leaflet.html', options=my_options, block=False)

So it seems to have unblocked for the time being, on the command line and also within the app itself ('block=False' is still within the code in the app but I took the my_options out, like so: eel.start('leaflet.html', block=False)).

I'm not sure what I'm looking at or what has occurred here, my intention is to package this app and distribute it so I can only imagine this issue will recur in quite a few other instances. Again any insight is appreciated.

**Comment by CentralHeating** [comment]
UPDATE:

I ran it from the command line using the following:
import eel
eel.init("")
eel.start('leaflet.html', block=False)

This also popped up a window that said 'localhost connection refused'. I then took out the 'block=False' and ran the same line again:
eel.start('leaflet.html')

Now it worked, it seems to have unblocked the page. The command line gave the following traceback when doing this:

-------------------------------------------------------------------------------------------------
Traceback (most recent call last):
File "src\gevent\greenlet.py", line 766, in gevent.greenlet.Greenlet.run
File "C:\Python27\lib\site-packages\eel_init.py", line 127, in run_lambda
quiet=True)
File "C:\Python27\lib\site-packages\bottle.py", line 3129, in run
server.run(app)
File "C:\Python27\lib\site-packages\bottle_websocket\server.py", line 17, in run
server.serve_forever()
File "C:\Python27\lib\site-packages\gevent\baseserver.py", line 367, in serve_forever
self.start()
File "C:\Python27\lib\site-packages\gevent\baseserver.py", line 305, in start
self.init_socket()
File "C:\Python27\lib\site-packages\gevent\pywsgi.py", line 1490, in init_socket
StreamServer.init_socket(self)
File "C:\Python27\lib\site-packages\gevent\server.py", line 146, in init_socket
self.socket = self.get_listener(self.address, self.backlog, self.family)
File "C:\Python27\lib\site-packages\gevent\server.py", line 157, in get_listener
return _tcp_listener(address, backlog=backlog, reuse_addr=cls.reuse_addr, family=family)
File "C:\Python27\lib\site-packages\gevent\server.py", line 256, in _tcp_listener
sock.bind(address)
File "", line 1, in bind
error: [Errno 10048] Only one usage of each socket address (protocol/network address/port) is normally permitt
ed: ('localhost', 8000)
2019-07-10T21:39:34Z <Greenlet at 0x30812d8: run_lambda> failed with error
---------------------------------------------------------------------------------------------------------

So it appears to be a port issue. In my original flailing trying to find an answer I also added options to the eel.start() function like so:

my_options = {
'mode': "chrome-app", #or "chrome-app",
'host': 'localhost',
'host': '127.0.0.1',
'port': 8080
}

eel.start('leaflet.html', options=my_options, block=False)

So it seems to have unblocked for the time being, on the command line and also within the app itself ('block=False' is still within the code in the app but I took the my_options out, like so: eel.start('leaflet.html', block=False)).

I'm not sure what I'm looking at or what has occurred here, my intention is to package this app and distribute it so I can only imagine this issue will recur in quite a few other instances. Again any insight is appreciated.

**Comment by CentralHeating** [comment]
Now it's not working again.

**Comment by mvincenzi96** [comment]
I am having exactly this same issue...

**Comment by ChrisKnott** [comment]
I think this is probably due to previous instances of the program not closing down cleanly. Check your task manager for running Python processes

**Comment by samuelhwilliams** [comment]
I'm going to close this based on inactivity after the previous response. Hopefully you got it working again. If you didn't, please feel free to comment again or open a new issue.

Thanks :)

**Comment by MeBadDev** [comment]
`killall python3` seems to fix it on my side, looks like it appears when you use `^C` to kill the last instance
